### ISSUE TITLE
[New Concept Docs]: Recursion in Python

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3052]
This issue describes how to implement the `recursion` concept docs.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

## Goal

This concept is meant to teach an understanding/use of `recursion`  in Python. 

<br>

## Learning objectives

- Understand what `recursion` means in general, and its use in programming.
- Understand the execution difference between loops and recursion (_e.g., one frame vs mulit-frame_).
- Understand how `recursion` can be used in Python (_anything done with looping can be done by recursing_)
   - Pros and cons of recursing
   - Problems that lend themselves to recursion vs problems that lend themselves to looping
   - Reasoning about code
   - Idempotence and single responsibility
- Understand the limits of a recursive approach in Python (_e.g. no tail-call optimization, Python's recursion limit, speed, etc_)
    - `sys.getrecursionlimit()` & `sys.setrecursionlimit()` -- and why you don't want to mess with things.
    - Various "mitigation" strategies for Pythons recursion limits.
    -  (Possibly?) `lru.cache`, `generators`, `comprehensions`, and `decorators` and their use in recursion strategies.
- (Possibly?) compare how Python implements recursion vs other lanugages.

<br>

## Out of scope

This is a fairly high-level and broad document, emphasizing a programming _approach_.  There might be one or more associated concept exercises, and perhaps multiple "Approaches" `instruction_append.md` docs across existing practice exercises.  As such, no topic is really "out of scope", as long as it applies to recursion, thinking recursively, the pros and cons of working with recursion, and how recursion differs from other approaches.

<br>
  
## Concepts

<details>
<summary><b>Proposed Concepts to Cover</b></summary>

- `recursion`
- `recursive thinking`
- `iteration` & `loops` (_by way of comparison_)
- `call stack`
- `memoization` (_as a recursion limit mitigation_)
- `decorators` (_as a recursion limit mitigation_)
-  `generators` (_as a recursion limit mitigation_)

</details>

<br>

## Prerequisites

Since this is a broad and cross-exercise concept, the prerequisites listed below are for any concept exercises that might be purpose-built for recursion.  Currently, we do not have any recursion concept exercises mapped out.  We'll need to revisit the prerequisites when we do map exercises, and will probably need to adjust them on an exercise-by-exercise basis.

<details>
<summary>Prereqs</summary>

- `basics`
- `booleans`
- `comparisons`
- `decorators`
- `dicts`
- `functions`
- `higher-order-functions`
- `iteration`
- `lists`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `tuples`

</details>

<br>  

## Resources to refer to

<details>
<summary><b>Articles on Recursion</b></summary>

- [Programiz: recursion](https://www.programiz.com/python-programming/recursion)
- [Real Python: recursion](https://realpython.com/python-recursion/)
- [Real Python: thinking recursively](https://realpython.com/python-thinking-recursively/)
- [Python Docs: `sys.setrecursionlimit()`](https://docs.python.org/3.8/library/sys.html#sys.setrecursionlimit)
- [Python Docs: `RecursionError`](https://docs.python.org/3.8/library/exceptions.html#RecursionError)
- [Towards Data Science: Stack Frames and Tail-Call Optimization](https://towardsdatascience.com/python-stack-frames-and-tail-call-optimization-4d0ea55b0542)
- [MIT EDU: Recursion](https://web.mit.edu/6.005/www/fa15/classes/10-recursion/)
- [After Academy: Analysis of Recursion in Programming](https://afteracademy.com/blog/analysis-of-recursion-in-programming)
- [Al Sweigart: Recursion for Beginners: A Beginners Guide to Recursion (Video)](https://www.youtube.com/watch?v=AfBqVVKg4GE)
- [Free Code Camp: How Recursion Works](https://www.freecodecamp.org/news/how-recursion-works-explained-with-flowcharts-and-a-video-de61f40cb7f9/)
- [Ned Batchelder: A Thing I Learned About Python Recursion](https://nedbatchelder.com/blog/201812/a_thing_i_learned_about_python_recursion.html)
- [Net Batchelder: Recursive Dogma](https://nedbatchelder.com/blog/201205/recursive_dogma.html)
- [Analytics India Mag: Ultimate Guide to Recursion and Iteration in Python](https://analyticsindiamag.com/ultimate-guide-to-recursion-and-iteration-in-python/)
- [Algo Daily: Problem Solving with Recursion vs Iteration](https://algodaily.com/lessons/problem-solving-with-recursion-vs-iteration)
- [Towards Data Science: Don't Use Recursion in Python Any More](https://towardsdatascience.com/dont-use-recursion-in-python-any-more-918aad95094c)

</details>
<details>
<summary><b>Different Recursive Problems, and Problem Ideas</b></summary>

- Ideas Based on Math and Sequences
   - [Towards Data Science: Visualized Riemann Integration in Python](https://towardsdatascience.com/integrals-are-easy-visualized-riemann-integration-in-python-87dd02e90994) (_AKA, estimating the area under a curve by using rectangles of decreasing size_)
   - [Math Libre Texts: Exponential Growth and Decay](https://math.libretexts.org/Bookshelves/Calculus/Book%3A_Calculus_(OpenStax)/06%3A_Applications_of_Integration/6.8%3A_Exponential_Growth_and_Decay) (_see also [Fox Rabbit math model](https://niko.roorda.nu/computer-programs/fox-rabbit-math-model/), Radioactive decay, Bacteria Growth, and Compounding Interest._)
   - [Turito:  Exponential Growth and Decay](https://www.turito.com/learn/math/exponential-growth-and-decay)
   - [Wikipedia:  Lucas Numbers](https://en.wikipedia.org/wiki/Lucas_number)
   - [Wikipedia: Wythoff Array](https://en.wikipedia.org/wiki/Wythoff_array)
   - [Wikipedia: Wythoff's Game](https://en.wikipedia.org/wiki/Wythoff%27s_game)
   - [Wikipedia: (The Game of) Nim](https://en.wikipedia.org/wiki/Nim) 
   - [Math Communities dot Org: Mondrain Art Puzzles](https://mathcommunities.org/mondrian-art-puzzles/) also [BJC EDC: Recursive Mondrain](https://bjc.edc.org/bjc-r/cur/programming/7-recursion/2-projects/4-mondrian.html?topic=nyc_bjc%2F7-recursion-trees-fractals.topic&course=bjc4nyc.html&novideo&noassignment)
- Existing Problems and Problem Lists from Different Sites
   - [Include Help: Egg Drop](https://www.includehelp.com/icp/egg-dropping-problem.aspx)
   - [Include Help: Weighted Job Scheduling](https://www.includehelp.com/icp/weighted-job-scheduling.aspx)
   - [Include Help: Island Probability](https://www.includehelp.com/icp/island-probability.aspx)
   - [Include Help: List of Recursion Problems](https://www.includehelp.com/icp/recursion-practice-problems-with-solutions.aspx) (_many of these are variations or repeats of existing exercism practice problems_)
   - [Techie Delight:  All Shortest Routes in a Rectangular Grid](https://www.techiedelight.com/print-all-shortest-routes-rectangular-grid/)
   - [Techie Delight:  Find all Paths from the 1st Cell to the Last Cell in a Matrix](https://www.techiedelight.com/find-all-paths-from-source-to-destination-in-matrix/)
   - [Techie Delight: Recursion Practice Problems](https://www.techiedelight.com/recursion-practice-problems-with-solutions/) (_many of these feel too mathabatory_)
   - [CodeChef: Philosophers Stone](https://www.codechef.com/problems/ALORA1)
   - [CodeChef: Connecting Soldiers](https://www.codechef.com/problems/NOKIA)
   - [CodeChef: Fit Squares in a Triangle](https://www.codechef.com/problems/TRISQ)
   - [CodeChef: Recursion List](https://www.codechef.com/tags/problems/recursion) (_like others, these are mostly math or repeats_)
   - [Leetcode: Parsing a Boolean Expression](https://leetcode.com/problems/parsing-a-boolean-expression/)
  - [GeeksforGeeks:  Recursive Practice Problems](https://www.geeksforgeeks.org/recursion-practice-problems-solutions/) (_not great, but included for completeness_)

</details>

<br>

## Files to Be Created

Please see the following for more details on these files:  [concepts](https://exercism.org/docs/building/tracks/concepts)

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

* ### Concept `about.md`

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-aboutmd)

  * This can also be a summary/paraphrase of the  `about.md` document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the associated concept or practice exercises.   It should contain a good summation of the concept, but not go into lots of detail.

<br>

## Implementation Notes

- Example code  should **only use syntax & concepts introduced within these docs or one of the prerequisite concept exercises or documents.**
  Please **do not use** syntax not previously covered in prerequisite topics or exercises.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

**Comment by bobahop** [comment]
Already started!

**Comment by bobahop** [comment]
Addressed by the merging of #3051
